src/transformStateWithClones.js

```diff
@@ -7,7 +7,22 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  let newState = { ...state };
+  const states = [];
+
+  for (const key of actions) {
+    if (key.type === 'addProperties') {
+      Object.assign(newState, key.extraData);
+    } else if (key.type === 'removeProperties') {
+      for (const removeKey of key.keysToRemove) {
```

Comment:
Ensure that `key.keysToRemove` is an array when `key.type` is 'removeProperties'. If `keysToRemove` is not guaranteed to be an array, consider adding a validation step.